### PR TITLE
Allow HOST has no port

### LINKED ISSUE #22274: Version 1.18 crashes for me due to mailer.HOST config
### Description

Hello,

In a loop, 1.18 keeps crashing forever:

```
Server listening on :: port 22.
Server listening on 0.0.0.0 port 22.
2022/12/30 05:11:07 cmd/web.go:106:runWeb() [I] Starting Gitea on PID: 15
2022/12/30 05:11:07 ...s/setting/setting.go:613:deprecatedSetting() [E] Deprecated fallback `[server]` `LFS_CONTENT_PATH` present. Use `[lfs]` `PATH` instead. This fallback will be removed in v1.19.0
2022/12/30 05:11:07 cmd/web.go:160:runWeb() [I] Global init
2022/12/30 05:11:07 ...s/setting/setting.go:613:deprecatedSetting() [E] Deprecated fallback `[server]` `LFS_CONTENT_PATH` present. Use `[lfs]` `PATH` instead. This fallback will be removed in v1.19.0
2022/12/30 05:11:07 routers/init.go:116:GlobalInitInstalled() [I] Git Version: 2.36.3, Wire Protocol Version 2 Enabled (home: /data/gitea/home)
2022/12/30 05:11:07 routers/init.go:117:GlobalInitInstalled() [I] AppPath: /usr/local/bin/gitea
2022/12/30 05:11:07 routers/init.go:118:GlobalInitInstalled() [I] AppWorkPath: /app/gitea
2022/12/30 05:11:07 routers/init.go:119:GlobalInitInstalled() [I] Custom path: /data/gitea
2022/12/30 05:11:07 routers/init.go:120:GlobalInitInstalled() [I] Log path: /data/gitea/log
2022/12/30 05:11:07 routers/init.go:121:GlobalInitInstalled() [I] Configuration file: /data/gitea/conf/app.ini
2022/12/30 05:11:07 routers/init.go:122:GlobalInitInstalled() [I] Run Mode: Prod
Received signal 15; terminating.
```

Once I downgrade to 1.17, all is working again.

I'm not sure what to do, or how to investigate further.

MySQL is running fine:

```
2022-12-30T05:10:29.580620Z 0 [System] [MY-010931] [Server] /usr/sbin/mysqld: ready for connections. Version: '8.0.31'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server - GPL.
```

Also, I'm not sure how to fix the `LFS_CONTENT_PATH` issue.
What I currently have is:

```
LFS_START_SERVER = true
LFS_CONTENT_PATH = /data/git/lfs
LFS_JWT_SECRET   = ...
```

Do I just move the `LFS_CONTENT_PATH` like this?

```
[lfs]
PATH = /data/git/lfs
```

and leave the other `LFS_*` as they are?

Thank you very much.

### Gitea Version

1.18

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

Bundled in Docker

### Operating System

Bundled in Docker

### How are you running Gitea?

```
services:
  server:
    container_name: gitea_server
    image: gitea/gitea:latest
    environment:
      - USER_UID=...
      - USER_GID=...
      - DB_TYPE=mysql
      - DB_HOST=db:3306
      - RUN_MODE=prod
      - ROOT_URL="https://..../"
    env_file:
      - ./db.env
    restart: always
    networks:
      - gitea
    volumes:
      - /.../data:/data
    ports:
      - "127.0.0.1:...:3000"
    depends_on:
      - db
  db:
    container_name: gitea_db
    image: mysql:8.0
    restart: always
    env_file:
      - ./db.env
    networks:
      - gitea
    volumes:
      - /.../mysql:/var/lib/mysql
    cap_add:
      # mbind: Operation not permitted
      - SYS_NICE

networks:
  gitea:
    external: false
```

### Database

MySQL

### PR DESCRIPTION
**Pull request by lunny**
Fix #22274

This PR will allow `HOST` without port. Then a default port will be given in future steps.

### CODE REVIEW

File: modules/setting/mailer.go [hunk 1]
```diff
@@ -71,7 +71,9 @@ func newMailService() {
 		givenHost := sec.Key("HOST").String()
 		addr, port, err := net.SplitHostPort(givenHost)
 		if err != nil {
-			log.Fatal("Invalid mailer.HOST (%s): %v", givenHost, err)
+			if err.Error() != "missing port in address" {
+				log.Fatal("Invalid mailer.HOST (%s): %v", givenHost, err)
+			}
```
Comment (delvh) [inline modules/setting/mailer.go]: We don't need the double nesting, you can `&&` concatenate the conditions.
(I would do it if GitHub would be able to fix their suggestion algorithm)

Comment (delvh) [reply modules/setting/mailer.go]: Also, does this really fix anything?
To me, this looks, like it simply masks a fatal error and transforms it into a runtime error whenever a mail should be sent, as they don't have a port to send it to anymore.

Comment (lunny) [reply modules/setting/mailer.go]: > Also, does this really fix anything? To me, this looks, like it simply masks a fatal error and transforms it into a runtime error whenever a mail should be sent, as they don't have a port to send it to anymore.

When there is no port, later lines will give it a default port.

**Review by wolfogre — APPROVED** [review]

File: modules/setting/mailer.go [hunk 2]
```diff
@@ -71,7 +71,9 @@ func newMailService() {
 		givenHost := sec.Key("HOST").String()
 		addr, port, err := net.SplitHostPort(givenHost)
 		if err != nil {
-			log.Fatal("Invalid mailer.HOST (%s): %v", givenHost, err)
+			if err.Error() != "missing port in address" {
```
Comment (wxiaoguang) [inline modules/setting/mailer.go]: If `err != nil`, then `addr == ""`, then how does `sec.Key("SMTP_ADDR").MustString(addr)` work?

File: modules/setting/mailer.go [hunk 3]
```diff
@@ -71,7 +71,9 @@ func newMailService() {
...
+				log.Fatal("Invalid mailer.HOST (%s): %v", givenHost, err)
```
Comment (yardenshoham) [inline modules/setting/mailer.go]: Seeing as `mailer.HOST` is deprecated, shouldn't this log something like `mailer.SMTP_ADDR`?

Comment (lafriks) [reply modules/setting/mailer.go]: But error is about host setting being invalid, this will be removed when we drop HOST support

Comment (yardenshoham) [reply modules/setting/mailer.go]: Gotcha

**Comment by lunny** [comment]
@delvh @wxiaoguang done.

**Comment by wxiaoguang** [comment]
I do not understand how could this change be right. Leave the review to others.

**Review by delvh — APPROVED** [review]

File: modules/setting/mailer.go [hunk 4]
```diff
@@ -70,9 +70,12 @@ func newMailService() {
 	if sec.HasKey("HOST") && !sec.HasKey("SMTP_ADDR") {
 		givenHost := sec.Key("HOST").String()
 		addr, port, err := net.SplitHostPort(givenHost)
-		if err != nil {
+		if err != nil && err.Error() != "missing port in address" {
 			log.Fatal("Invalid mailer.HOST (%s): %v", givenHost, err)
 		}
+		if addr == "" {
+			addr = "127.0.0.1"
+		}
```
Comment (zeripath) [inline modules/setting/mailer.go]: If an error is returned then addr and port will be empty.

Therefore when there is an error we should re-set addr to `givenHost`.


```suggestion
		if err != nil && err.Error() == "missing port in address" {
			addr = givenHost   
		} else if err != nil {
			log.Fatal("Invalid mailer.HOST (%s): %v", givenHost, err)
		}
		if addr == "" {
			addr = "127.0.0.1"
		}		
```

I don't think it's necessarily correct to re-set an empty SMTP_ADDR to 127.0.0.1 but I guess it's not necessarily incorrect

**Review by zeripath — CHANGES_REQUESTED** [review]

**Comment by zeripath** [comment]
> Are you really sure this PR is right?
> 
> Suppose `HOST = my.domain`
> 
> Then you get `SMTP_ADDR = 127.0.0.1` in the end. Won't it mislead users seriously?

@wxiaoguang take another look at the PR after my requested changes.

**Comment by zeripath** [comment]
@lunny unfortunately I've messed up the whitespace in my suggestion so it's gonna need a `make fmt`

**Comment by lunny** [comment]
> @lunny unfortunately I've messed up the whitespace in my suggestion so it's gonna need a `make fmt`

Done.

**Review by zeripath — APPROVED** [review]

**Comment by lunny** [comment]
Please review again, I added some changes with some tests.

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
Please send backport.

**Pull request merged**